core/src/fixed_window.rs

```diff
@@ -0,0 +1,206 @@
+use std::borrow::Borrow;
+use std::cmp::Eq;
+use std::collections::hash_map::Entry::{Occupied, Vacant};
+use std::collections::HashMap;
+use std::hash::Hash;
+use std::ops::AddAssign;
+use std::time::{Duration, Instant};
+
+/// A fixed lookback window maintains some statistics over the half-open time
+/// interval:
+///    [start  <- duration ->  end)
+/// When time passes the end of the window, a new window is instantiated.
+/// Compared to "rolling" windows, it has the advantage of computational and
+/// memory efficiency. However, at the beginning of each window, it starts from
+/// scratch and has zero data. To mitigate this, this implementation uses two
+/// windows simultaneously, one starting half the interval after the other.
+/// When looking up the values, it uses the interval which starts earlier.
+pub struct FixedWindow<K, V> {
+    /// Length of lookback window.
+    duration: Duration,
+    /// Beginning of each lookback window.
+    /// One should start half the duration after the other.
+    epoch: [Instant; 2],
+    table: [HashMap<K, V>; 2],
+}
+
+impl<K, V> FixedWindow<K, V>
+where
+    K: Eq + Hash + Clone,
+{
+    pub fn new(now: Instant, duration: Duration) -> Self {
+        FixedWindow {
+            duration,
+            epoch: [now, now + duration / 2],
+            table: [HashMap::new(), HashMap::new()],
+        }
+    }
+
+    /// Looks up the key from the lookback window.
+    /// Returns None if the key does not exist or if both windows are expired.
+    pub fn get<Q: ?Sized>(&self, now: Instant, key: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        // If both windows are expired, i.e.
+        //   self.epoch[i] + self.duration <= now
+        // then return None.
+        // If exactly one of the two windows is expired, then look up from the
+        // other window. Otherwise, pick the window which starts earlier.
+        if self.epoch[0] + self.duration <= now {
+            if self.epoch[1] + self.duration <= now {
+                None
+            } else {
+                self.table[1].get(key)
+            }
+        } else if self.epoch[1] + self.duration <= now || self.epoch[0] < self.epoch[1] {
+            self.table[0].get(key)
+        } else {
+            self.table[1].get(key)
+        }
+    }
+
+    /// Records a new value for the lookback window.
+    pub fn add<I>(&mut self, now: Instant, key: K, item: I)
+    where
+        V: From<I> + AddAssign<I>,
+        I: Clone,
+    {
+        for i in 0..2 {
+            // If the window is expired, discard the values and restart.
+            if self.epoch[i] + self.duration <= now {
+                self.epoch[i] = Instant::max(now, self.epoch[1 - i] + self.duration / 2);
+                self.table[i].clear();
+            }
+            if self.epoch[i] <= now {
+                match self.table[i].entry(key.clone()) {
```

Comment:
I think this can be simplified to:

```
self.table[i]
    .entry(key.clone())
    .and_modify(|e| *e += item.clone())
    .or_insert(item.clone().into());
```

Reply:
done! thanks that is indeed simpler